lib/Dialect/HW/HWTypes.cpp

```diff
@@ -144,6 +144,16 @@ bool circt::hw::hasHWInOutType(Type type) {
   return type.isa<InOutType>();
 }
 
+ArrayType circt::hw::castArrayType(Type type) {
+  if (auto arrayType = type.dyn_cast<ArrayType>())
+    return arrayType;
+
+  if (auto aliasType = type.dyn_cast<TypeAliasType>())
+    return aliasType.getCanonicalType().cast<ArrayType>();
```

Comment:
I think `cast` will assert on fail, which you don't want.

Reply:
We know statitically that if it is a type alias, then its canonical type is an array type. I was thinking about this, and I think I will just make TypeAliasOr a proper TypeDef, so this casting can be a generic `getCanonicalType` on `TypeAliasOr`, rather than having this (and other) free functions to do the cast.

Reply:
Actually, I don't think subclassing TypeDef is what we want... that would add another layer of wrapping the type during parsing and printing, when really we just want to represent that it can parse/print as either an alias or the inner type, and have a method to canonicalize to the inner type when neccessary. So what we really want is a DialectType, with an extra helper method for casting... that's basically what this is. I don't know if there is a way to do this and attach the helper to the DialectType rather than just using a free function like this.

Reply:
For now I've made it a DialectType and moved `cast`->`dyn_cast` so this returns null rather than asserting in case the assumption is not met. I also copied the comment about this behavior to the implementation file as well to make it clear.

I'm still curious if it would be possible to attach the `castArrayType` functionality onto the DialectType itself, but I think the only way to do that would be to add a new Type subclass and use it for the DialectType. I can explore that, but it may be more trouble and indirection than its worth.

Reply:
> I can explore that

It's not actually that bad... I will update this PR